src/vipersci/carto/heatmap.py

```diff
@@ -158,25 +166,38 @@ def generate_density_heatmap(
             center location and *radius*) that overlap a given grid cell.
         avg: Averaged value of all the individual observations that overlap
             a given grid cell.
+        frequencies: Can be provided to subsequent generate_density_heatmap calls
+            on new data sampled at the same locations
 
     The *counts* and *avg* objects are numpy arrays.
     """
 
+    def as_ndarray(input: Sequence) -> np.ndarray:
```

Comment:
If abused, this function could return `None`.  We might as well also use Numpy Typing here an elsewhere, how about:
```
from numpy.typing import NDArray  # at top of file, of course.

if isinstance(input, NDArray):
    return input
else:
    return np.asarray(input)
````

Reply:
`numpy.typing.NDArray` is actually a generic type defined as `numpy.ndarray[typing.Any, numpy.dtype[+ScalarType]]`.  As it turns out `isinstance()` doesn't like parametrized generics, so I think we're stuck with `np.ndarray` here.

It makes sense to use NDArray in all of the type annotions, however.  I have changed them.

Reply:
Whoops my bad on the isinstance(), you're right.